File: packages/core/src/stats/regression.ts
/**
 * Fits and summaries that turn a scatter into a claim: a trend line with a
 * confidence band, a local-regression smoother, an ECDF, a correlation matrix.
 * Pure — every function takes arrays and returns arrays.
 */

/** An ordinary-least-squares fit of `y = slope·x + intercept`. */
export interface LinearFit {
  slope: number;
  intercept: number;
  /** Coefficient of determination, 0..1. */
  r2: number;
  /** Standard error of the residuals. */
  stderr: number;
  /** Points used (non-finite pairs are skipped). */
  n: number;
  /** Evaluate the fit at an x. */
  predict(x: number): number;
}

/** Least-squares straight line through (x, y), skipping non-finite pairs. */
export function linearRegression(x: ArrayLike<number>, y: ArrayLike<number>): LinearFit {
  const len = Math.min(x.length, y.length);
  let n = 0;
  let sx = 0;
  let sy = 0;
  for (let i = 0; i < len; i++) {
    const xi = x[i]!;
    const yi = y[i]!;
    if (!Number.isFinite(xi) || !Number.isFinite(yi)) continue;
    sx += xi;
    sy += yi;
    n++;
  }
  if (n < 2) {
    const intercept = n === 1 ? sy : 0;
    return { slope: 0, intercept, r2: 0, stderr: 0, n, predict: () => intercept };
  }
  const mx = sx / n;
  const my = sy / n;
  let sxx = 0;
  let sxy = 0;
  let syy = 0;
  for (let i = 0; i < len; i++) {
    const xi = x[i]!;
    const yi = y[i]!;
    if (!Number.isFinite(xi) || !Number.isFinite(yi)) continue;
    const dx = xi - mx;
    const dy = yi - my;
    sxx += dx * dx;
    sxy += dx * dy;
    syy += dy * dy;
  }
  const slope = sxx === 0 ? 0 : sxy / sxx;
  const intercept = my - slope * mx;
  const ssRes = Math.max(0, syy - slope * sxy);
  const r2 = syy === 0 ? 1 : 1 - ssRes / syy;
  const stderr = n > 2 ? Math.sqrt(ssRes / (n - 2)) : 0;
  return { slope, intercept, r2, stderr, n, predict: (v: number) => slope * v + intercept };
}

/** A smoothed curve sampled on a grid, optionally with a band around it. */
export interface Trend {
  x: Float64Array;
  y: Float64Array;
  /** Lower/upper band, when the caller asked for one. */
  lower?: Float64Array;
  upper?: Float64Array;
}

/**
 * Sample a linear fit across the data range, with an optional ±`k`·stderr band.
 * `points` controls the resolution (2 is enough for the line itself).
 */
export function linearTrend(
  x: ArrayLike<number>,
  y: ArrayLike<number>,
  opts: { points?: number; band?: number } = {},
): Trend {
  const fit = linearRegression(x, y);
  const points = Math.max(2, opts.points ?? 2);
  let lo = Infinity;
  let hi = -Infinity;
  for (let i = 0; i < x.length; i++) {
    const v = x[i]!;
    if (!Number.isFinite(v)) continue;
    if (v < lo) lo = v;
    if (v > hi) hi = v;
  }
  if (!Number.isFinite(lo)) { lo = 0; hi = 1; }
  const gx = new Float64Array(points);
  const gy = new Float64Array(points);
  for (let i = 0; i < points; i++) {
    gx[i] = lo + ((hi - lo) * i) / (points - 1);
    gy[i] = fit.predict(gx[i]!);
  }
  if (!opts.band) return { x: gx, y: gy };
  const k = opts.band;
  const lower = new Float64Array(points);
  const upper = new Float64Array(points);
  for (let i = 0; i < points; i++) {
    lower[i] = gy[i]! - k * fit.stderr;
    upper[i] = gy[i]! + k * fit.stderr;
  }
  return { x: gx, y: gy, lower, upper };
}

/**
 * LOESS: locally-weighted linear regression with a tricube kernel. `bandwidth`
 * is the fraction of points in each local neighbourhood (0..1, default 0.3) —
 * larger is smoother. Returns the fit sampled on a sorted grid, so it can be
 * drawn straight as a line.
 */
export function loess(
  x: ArrayLike<number>,
  y: ArrayLike<number>,
  opts: { bandwidth?: number; points?: number } = {},
): Trend {
  const len = Math.min(x.length, y.length);
  const pairs: Array<[number, number]> = [];
  for (let i = 0; i < len; i++) {
    const xi = x[i]!;
    const yi = y[i]!;
    if (Number.isFinite(xi) && Number.isFinite(yi)) pairs.push([xi, yi]);
  }
  pairs.sort((a, b) => a[0] - b[0]);
  const n = pairs.length;
  const grid = Math.max(2, Math.min(opts.points ?? 100, Math.max(2, n)));
  const gx = new Float64Array(grid);
  const gy = new Float64Array(grid);
  if (n === 0) return { x: gx, y: gy };

  const bandwidth = Math.min(1, Math.max(2 / n, opts.bandwidth ?? 0.3));
  const span = Math.max(2, Math.round(bandwidth * n));
  const lo = pairs[0]![0];
  const hi = pairs[n - 1]![0];

  for (let g = 0; g < grid; g++) {
    const x0 = grid === 1 ? lo : lo + ((hi - lo) * g) / (grid - 1);
    gx[g] = x0;
    // The `span` nearest neighbours set the local scale.
    const dists = pairs.map(([px], i) => [Math.abs(px - x0), i] as const);
    dists.sort((a, b) => a[0] - b[0]);
    const maxDist = dists[Math.min(span, n) - 1]![0] || 1e-12;
    let sw = 0, swx = 0, swy = 0, swxx = 0, swxy = 0;
    for (let k = 0; k < Math.min(span, n); k++) {
      const [d, idx] = dists[k]!;
      const u = d / maxDist;
      const w = u >= 1 ? 0 : (1 - u * u * u) ** 3; // tricube
      if (w === 0) continue;
      const [px, py] = pairs[idx]!;
      sw += w;
      swx += w * px;
      swy += w * py;
      swxx += w * px * px;
      swxy += w * px * py;
    }
    if (sw === 0) { gy[g] = pairs[Math.min(n - 1, g)]![1]; continue; }
    const denom = sw * swxx - swx * swx;
    if (Math.abs(denom) < 1e-12) { gy[g] = swy / sw; continue; }
    const slope = (sw * swxy - swx * swy) / denom;
    const intercept = (swy - slope * swx) / sw;
    gy[g] = slope * x0 + intercept;
  }
  return { x: gx, y: gy };
}

/** The empirical CDF as a step function: sorted values against their cumulative proportion. */
export function ecdf(values: ArrayLike<number>): { x: Float64Array; y: Float64Array } {
  const xs: number[] = [];
  for (let i = 0; i < values.length; i++) if (Number.isFinite(values[i]!)) xs.push(values[i]!);
  xs.sort((a, b) => a - b);
  const n = xs.length;
  const x = new Float64Array(n);
  const y = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    x[i] = xs[i]!;
    y[i] = (i + 1) / n;
  }
  return { x, y };
}

/** Standardize to zero mean and unit variance (population sd). Non-finite entries pass through. */
export function zscore(values: ArrayLike<number>): Float64Array {
  const n = values.length;
  const out = new Float64Array(n);
  let count = 0;
  let mean = 0;
  for (let i = 0; i < n; i++) if (Number.isFinite(values[i]!)) { mean += values[i]!; count++; }
  if (count === 0) return out;
  mean /= count;
  let varSum = 0;
  for (let i = 0; i < n; i++) if (Number.isFinite(values[i]!)) varSum += (values[i]! - mean) ** 2;
  const sd = Math.sqrt(varSum / count) || 1;
  for (let i = 0; i < n; i++) out[i] = Number.isFinite(values[i]!) ? (values[i]! - mean) / sd : values[i]!;
  return out;
}

/** Pearson correlation of two equal-length series (0 when either is constant). */
export function correlation(a: ArrayLike<number>, b: ArrayLike<number>): number {
  const n = Math.min(a.length, b.length);
  let ma = 0, mb = 0, count = 0;
  for (let i = 0; i < n; i++) {
    if (!Number.isFinite(a[i]!) || !Number.isFinite(b[i]!)) continue;
    ma += a[i]!; mb += b[i]!; count++;
  }
  if (count < 2) return 0;
  ma /= count; mb /= count;
  let saa = 0, sbb = 0, sab = 0;
  for (let i = 0; i < n; i++) {
    if (!Number.isFinite(a[i]!) || !Number.isFinite(b[i]!)) continue;
    const da = a[i]! - ma;
    const db = b[i]! - mb;
    saa += da * da; sbb += db * db; sab += da * db;
  }
  const denom = Math.sqrt(saa * sbb);
  return denom === 0 ? 0 : sab / denom;
}

/**
 * Pearson correlation matrix of `columns`, row-major and `k×k`. Feed it straight
 * to `addHeatmap` with a diverging colormap and `symmetricDomain`.
 */
export function corrMatrix(columns: ReadonlyArray<ArrayLike<number>>): {
  values: Float64Array;
  size: number;
} {
  const k = columns.length;
  const values = new Float64Array(k * k);
  for (let i = 0; i < k; i++) {
    values[i * k + i] = 1;
    for (let j = i + 1; j < k; j++) {
      const r = correlation(columns[i]!, columns[j]!);
      values[i * k + j] = r;
      values[j * k + i] = r;
    }
  }
  return { values, size: k };
}

File: native/src/stats/regression.cpp
#include "stats/regression.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <utility>

namespace photon::stats {

LinearFit linear_regression(const double* x, const double* y, size_t count) {
  LinearFit fit;
  if (!x || !y) return fit;
  double sx = 0.0;
  double sy = 0.0;
  for (size_t i = 0; i < count; ++i) {
    if (!std::isfinite(x[i]) || !std::isfinite(y[i])) continue;
    sx += x[i];
    sy += y[i];
    ++fit.n;
  }
  if (fit.n < 2) {
    // One point is a horizontal line through it; none is a line through zero.
    fit.intercept = fit.n == 1 ? sy : 0.0;
    return fit;
  }
  const double dn = static_cast<double>(fit.n);
  const double mx = sx / dn;
  const double my = sy / dn;
  double sxx = 0.0;
  double sxy = 0.0;
  double syy = 0.0;
  for (size_t i = 0; i < count; ++i) {
    if (!std::isfinite(x[i]) || !std::isfinite(y[i])) continue;
    const double dx = x[i] - mx;
    const double dy = y[i] - my;
    sxx += dx * dx;
    sxy += dx * dy;
    syy += dy * dy;
  }
  fit.slope = sxx == 0.0 ? 0.0 : sxy / sxx;
  fit.intercept = my - fit.slope * mx;
  const double ss_res = std::max(0.0, syy - fit.slope * sxy);
  fit.r2 = syy == 0.0 ? 1.0 : 1.0 - ss_res / syy;
  fit.stderror = fit.n > 2 ? std::sqrt(ss_res / (dn - 2.0)) : 0.0;
  return fit;
}

Trend linear_trend(const double* x, const double* y, size_t count, int points, double band) {
  const LinearFit fit = linear_regression(x, y, count);
  const size_t n = static_cast<size_t>(std::max(2, points));
  double lo = std::numeric_limits<double>::infinity();
  double hi = -lo;
  for (size_t i = 0; x && i < count; ++i) {
    if (!std::isfinite(x[i])) continue;
    lo = std::min(lo, x[i]);
    hi = std::max(hi, x[i]);
  }
  if (!std::isfinite(lo)) {
    lo = 0.0;
    hi = 1.0;
  }
  Trend out;
  out.x.resize(n);
  out.y.resize(n);
  for (size_t i = 0; i < n; ++i) {
    out.x[i] = lo + (hi - lo) * static_cast<double>(i) / static_cast<double>(n - 1);
    out.y[i] = fit.predict(out.x[i]);
  }
  if (band == 0.0) return out;
  out.lower.resize(n);
  out.upper.resize(n);
  for (size_t i = 0; i < n; ++i) {
    out.lower[i] = out.y[i] - band * fit.stderror;
    out.upper[i] = out.y[i] + band * fit.stderror;
  }
  return out;
}

Trend loess(const double* x, const double* y, size_t count, double bandwidth, int points) {
  std::vector<std::pair<double, double>> pairs;
  if (x && y) {
    pairs.reserve(count);
    for (size_t i = 0; i < count; ++i) {
      if (std::isfinite(x[i]) && std::isfinite(y[i])) pairs.emplace_back(x[i], y[i]);
    }
  }
  std::sort(pairs.begin(), pairs.end(),
            [](const auto& a, const auto& b) { return a.first < b.first; });
  const size_t n = pairs.size();
  const size_t grid = static_cast<size_t>(
      std::max(2, std::min(points > 0 ? points : 100, std::max(2, static_cast<int>(n)))));
  Trend out;
  out.x.assign(grid, 0.0);
  out.y.assign(grid, 0.0);
  if (n == 0) return out;

  const double dn = static_cast<double>(n);
  const double frac =
      std::min(1.0, std::max(2.0 / dn, bandwidth > 0.0 ? bandwidth : 0.3));
  const size_t span = std::max<size_t>(2, static_cast<size_t>(std::lround(frac * dn)));
  const size_t take = std::min(span, n);
  const double lo = pairs.front().first;
  const double hi = pairs.back().first;

  // Reused across grid points so the loop allocates nothing.
  std::vector<std::pair<double, size_t>> dists(n);
  for (size_t g = 0; g < grid; ++g) {
    const double x0 = lo + (hi - lo) * static_cast<double>(g) / static_cast<double>(grid - 1);
    out.x[g] = x0;
    for (size_t i = 0; i < n; ++i) dists[i] = {std::abs(pairs[i].first - x0), i};
    // Only the nearest `take` matter, and their order among themselves does
    // not — but the TypeScript sorts the whole array, and the k-th distance it
    // picks has to be the same one, so a partial sort by the same comparator is.
    std::partial_sort(dists.begin(), dists.begin() + static_cast<ptrdiff_t>(take), dists.end(),
                      [](const auto& a, const auto& b) { return a.first < b.first; });
    double max_dist = dists[take - 1].first;
    if (max_dist == 0.0) max_dist = 1e-12;

    double sw = 0.0, swx = 0.0, swy = 0.0, swxx = 0.0, swxy = 0.0;
    for (size_t k = 0; k < take; ++k) {
      const double u = dists[k].first / max_dist;
      if (u >= 1.0) continue;
      const double t = 1.0 - u * u * u;
      const double w = t * t * t;  // tricube
      if (w == 0.0) continue;
      const auto& p = pairs[dists[k].second];
      sw += w;
      swx += w * p.first;
      swy += w * p.second;
      swxx += w * p.first * p.first;
      swxy += w * p.first * p.second;
    }
    if (sw == 0.0) {
      out.y[g] = pairs[std::min(n - 1, g)].second;
      continue;
    }
    const double denom = sw * swxx - swx * swx;
    if (std::abs(denom) < 1e-12) {
      out.y[g] = swy / sw;
      continue;
    }
    const double slope = (sw * swxy - swx * swy) / denom;
    const double intercept = (swy - slope * swx) / sw;
    out.y[g] = slope * x0 + intercept;
  }
  return out;
}

Trend ecdf(const double* values, size_t count) {
  Trend out;
  if (!values) return out;
  std::vector<double> xs;
  xs.reserve(count);
  for (size_t i = 0; i < count; ++i) {
    if (std::isfinite(values[i])) xs.push_back(values[i]);
  }
  std::sort(xs.begin(), xs.end());
  const size_t n = xs.size();
  out.x.resize(n);
  out.y.resize(n);
  for (size_t i = 0; i < n; ++i) {
    out.x[i] = xs[i];
    out.y[i] = static_cast<double>(i + 1) / static_cast<double>(n);
  }
  return out;
}

std::vector<double> zscore(const double* values, size_t count) {
  std::vector<double> out(count, 0.0);
  if (!values) return out;
  size_t used = 0;
  double mean = 0.0;
  for (size_t i = 0; i < count; ++i) {
    if (!std::isfinite(values[i])) continue;
    mean += values[i];
    ++used;
  }
  if (used == 0) return out;
  mean /= static_cast<double>(used);
  double var_sum = 0.0;
  for (size_t i = 0; i < count; ++i) {
    if (!std::isfinite(values[i])) continue;
    const double d = values[i] - mean;
    var_sum += d * d;
  }
  double sd = std::sqrt(var_sum / static_cast<double>(used));
  if (sd == 0.0) sd = 1.0;
  for (size_t i = 0; i < count; ++i) {
    out[i] = std::isfinite(values[i]) ? (values[i] - mean) / sd : values[i];
  }
  return out;
}

double correlation(const double* a, const double* b, size_t count) {
  if (!a || !b) return 0.0;
  double ma = 0.0;
  double mb = 0.0;
  size_t used = 0;
  for (size_t i = 0; i < count; ++i) {
    if (!std::isfinite(a[i]) || !std::isfinite(b[i])) continue;
    ma += a[i];
    mb += b[i];
    ++used;
  }
  if (used < 2) return 0.0;
  ma /= static_cast<double>(used);
  mb /= static_cast<double>(used);
  double saa = 0.0, sbb = 0.0, sab = 0.0;
  for (size_t i = 0; i < count; ++i) {
    if (!std::isfinite(a[i]) || !std::isfinite(b[i])) continue;
    const double da = a[i] - ma;
    const double db = b[i] - mb;
    saa += da * da;
    sbb += db * db;
    sab += da * db;
  }
  const double denom = std::sqrt(saa * sbb);
  return denom == 0.0 ? 0.0 : sab / denom;
}

std::vector<double> corr_matrix(const double* const* columns, size_t k, size_t count) {
  std::vector<double> out(k * k, 0.0);
  if (!columns) return out;
  for (size_t i = 0; i < k; ++i) {
    out[i * k + i] = 1.0;
    for (size_t j = i + 1; j < k; ++j) {
      const double r = correlation(columns[i], columns[j], count);
      out[i * k + j] = r;
      out[j * k + i] = r;
    }
  }
  return out;
}

}  // namespace photon::stats
